url/LocalRouter.ts

```diff
@@ -0,0 +1,228 @@
+        }
+    }
+
+    return result;
+}
+
+/**
+ * Generates the query string
+ */
+function generateQueryString (route: string, params: RouteParams) : string
+{
+    return "?" + stringify(extend({route}, params), {
+        skipNull: true,
+        skipEmptyString: true,
+        // sort parameters, so that "route" is always the first one
+        sort: (left, right) =>
+        {
+            if (left === "route")
+            {
+                return -1;
+            }
+
+            if (right === "route")
+            {
+                return 1;
+            }
+
+            if (left === right)
+            {
+                return 0;
+            }
+
+            return left < right ? -1 : 1;
+        },
+    });
+}
+
+/**
+ * Parses the current query parameters
+ */
+function parseQueryString (queryString: string) : RouteParams
+{
+    return parse(queryString, {parseNumbers: true}) as RouteParams;
+}
+
+function paramsAreValid (params: RouteParams) : params is ValidRouteParams
+{
+    return typeof params.route === "string";
+}
+
+/**
+ * A local router that uses URL query parameters for routing
+ */
+export class LocalRouter
+{
+    private collection: Record<string, RouteDefinition<any>> = {};
+    private useUrl: boolean;
+
+
+    /**
+     */
+    public constructor (updateUrl: boolean = true)
+    {
+        this.useUrl = updateUrl;
+    }
+
+
+    /**
+     * Registers a new route handler
+     */
+    public on <TRouteParams> (
+        name: string,
+        handler: RouteHandler<TRouteParams>,
+        normalizers: RouteParamNormalizers<TRouteParams>
+    ) : this
+    {
+        if (this.collection[name])
+        {
+            throw new Error("Can't define two routes with the same name.");
+        }
+
+        if ("route" in normalizers)
+        {
+            throw new Error("Can't use 'route' as route parameter");
+        }
+
+        this.collection[name] = {handler, normalizers};
+
+        return this;
+    }
+
+
+    /**
+     * Initializes the router
+     */
+    public init (fallbackRoute: string, fallbackParams: RouteParams = {}) : void
+    {
+        if (this.useUrl)
+        {
+            // register event listener
+            on(window, "popstate", () => this.match(document.location.search));
+
+            // try to match
+            if (this.match(document.location.search))
+            {
+                return;
+            }
+        }
+
+        this.tryHandle(fallbackRoute, fallbackParams);
+    }
+
+
+    /**
+     * Navigates to the given route
+     */
+    public navigate (route: string, params: RouteParams = {}): void
+    {
+        if (!this.collection[route])
+        {
+            throw new Error(`Route not found: ${route}`);
+        }
+
+        // let the exception bubble through
+        const normalizedParameters = this.tryHandle(route, params);
+
+        if (this.useUrl && normalizedParameters)
```

Comment:
What is the use-case in ever wanting to set `this.useUrl` to anything but true? What is the expected behavior when I then try to invoke `navigate`?

Reply:
The expected use case already exists: we use the router in embedded CRUDs (for example to manage block entries), that are not a standalone-view. Updating the UI there wouldn't make sense.

The expected (and implemented) behaviour is that everything works **except** that the URL doesn't change.

This part is completely unchanged to the current (internal) `LocalRouter`.